### DISCUSSION TITLE
Allow googlebot through a paywall on a page using Incremental Static Regeneration

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#16665] [Help] [answered]
Described here [Make sure Google can crawl and index your pages](https://developers.google.com/search/docs/data-types/paywalled-content#googlebot) and here [Verifying Googlebot](https://support.google.com/webmasters/answer/80553) are the recommended ways of allowing Googlebot to index the paywalled content of a page.

Some of our page content (like section headings and a short summary) is available publicly, so this is fetched in a `getStaticProps` function, with a `revalidate` set. On the other hand, some of the content is only available to subscribed/authenticated users through a paywall, and fetched client-side after user authentication. We would like to let Googlebot index that content as well though. With SSR, we could check the incoming IP and user-agent to verify the requester, and let the bot load the whole page as if it was an authenticated user, but this cannot be done safely on client-side.

Is there any way to check the user-agent/IP in `getStaticPaths` and return different data to the client based on that?

**Comment by balazsorban44** [comment, marked as answer]
**Update**:
To answer my own question, what we ended up with was to check `navigator.userAgent` as a "soft" security measure, then make a request to the backend for the data behind authentication. Only then, in addition to the bearer token auth method, we also check the user agent and the corresponding request IP address as recommended by Google [here](https://support.google.com/webmasters/answer/80553).

This way, both authenticated users and Googlebot will be able to pass the paywall.


So to sum it up, this is what I do:
1. User opens the page
2. `getStaticProps` is invoked and returns the public data for the page. (In reality, it has already been invoked before, so the user doesn't actually wait for data here. See [data fetching](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation)
3. The user is presented with a page with partial (public) data, and a paywall on top of it. The paywall is closable so the user can read through the public content.
4. A `fetch` call for the paywalled (non-public) content is conditionally invoked, if either of the following conditions are met:
- The user's authentication status checks out, and has a valid access token (we are using https://next-auth.js.org/)
- The user's user agent is checked if it is Googlebot (using `navigator.userAgent`)
```js
const isGooglebot = typeof window === "undefined" ? false : navigator.userAgent.toLowerCase().includes("googlebot")
```
7. If the user is authenticated, or `isGooglebot === true`, the `fetch` call is made to the API endpoint for the non-public data:
- If the user had a valid `Bearer token`, it was sent with the request, and the API will allow the return of the paywalled data.
- If there was no access token, but the user agent contained Googlebot, an additional [reverse DNS lookup](https://support.google.com/webmasters/answer/80553) is run. If that also verifies, the user (in this case Google crawler) gets the data as well.
8. The client gets a response with the paywalled data, that is now merged with the public data.
9. The paywall is removed, and the page rerenders, now with the whole content.

**Reply by rshackleton** [reply]
@balazsorban44 could you please outline the solution in a little more detail? I'm looking into the best way to approach this and was running into the same challenges with balancing fast SSR output and secure paywalled content, whilst still allowing the content to be crawled by Google.

I'm not sure from your answer how the different requests are being handled and when, or how, you are checking the request is from Google.

How accurate is this summary of your approach?

- SSR the content without the paywalled section
- On client-side fetch serverless function:
    - In serverless function, validate the request (authentication/marketing opt-in cookie/Google DNS verification)
        - If not valid then return no content
    - If valid then return content
- Handle the API response:
    - If the content is returned, display it with structured data markup
    - If no content is returned then display the access form

**Reply by balazsorban44** [reply]
@rshackleton I updated my answer, I hope it helps. :slightly_smiling_face: 

No server-side rendering is happening, only SSG (static site generation) and a client-side fetch!

Please tell me, if it is still unclear!

**Reply by rshackleton** [reply]
Perfect, thanks for the detailed explanation as I'm sure others will happen upon this issue in the future. 🎉

**Reply by ahetawal-p** [reply]
@balazsorban44 stumbled upon this thread, while researching the same issue.
I have 2 questions for you, was wondering how the indexing results look 
1. With the client side fetch is the crawler still able to see the paywall content. My limited understanding is we always need SSR for indexing to work correctly. JS based or SPA based apps don't perform well.
2. In terms of actual paywall content, are you following the Google recommended structured data format to differentiate public vs paywall. Any example of how you do that?

Thanks again.
